[OData] Order By On Related Entities

The OData V2 module currently not supports order by statements on expanded entities. E.g when calling our service with

`GET <baseUrl>/UseCasePattern?$skip=0&$top=20&$expand=Style&$orderby=Style/Name` 

we get following error in stacktrace:

```
Caused by: org.apache.olingo.odata2.api.exception.ODataNotImplementedException: Not implemented
	at org.eclipse.dirigible.engine.odata2.sql.builder.SQLQuery.orderBy(SQLQuery.java:105) ~[dirigible-odata-core-5.12.9.jar:na]
	at org.eclipse.dirigible.engine.odata2.sql.builder.SQLQueryBuilder.buildSelectEntitySetIdsForTopAndExpandQuery(SQLQueryBuilder.java:135) ~[dirigible-odata-core-5.12.9.jar:na]
	at org.eclipse.dirigible.engine.odata2.sql.processor.AbstractSQLProcessor.readIdsForExpand(AbstractSQLProcessor.java:302) ~[dirigible-odata-core-5.12.9.jar:na]
	at org.eclipse.dirigible.engine.odata2.sql.processor.AbstractSQLProcessor.readEntitySet(AbstractSQLProcessor.java:230) ~[dirigible-odata-core-5.12.9.jar:na]
	at org.eclipse.dirigible.engine.odata2.sql.processor.AbstractSQLProcessor$$FastClassBySpringCGLIB$$f667424d.invoke(<generated>) ~[dirigible-odata-core-5.12.9.jar:na]
	at org.springframework.cglib.proxy.MethodProxy.invoke(MethodProxy.java:218) ~[spring-core-5.3.8.jar:5.3.8]
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:688) ~[spring-aop-5.3.8.jar:5.3.8]
```

Example call on OData V2 Testservice:

`GET https://services.odata.org/v2/OData/OData.svc/Products?$orderby=Rating,Category/Name desc`

For us it would be really helpful to get this feature.

--- a/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/builder/SQLQuery.java
+++ b/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/builder/SQLQuery.java
@@ -11,16 +11,59 @@
  */
 package org.eclipse.dirigible.engine.odata2.sql.builder;
 
+import static org.apache.olingo.odata2.api.commons.HttpStatusCodes.BAD_REQUEST;
+import static org.apache.olingo.odata2.api.commons.HttpStatusCodes.INTERNAL_SERVER_ERROR;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.FROM;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.INTO;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.JOIN;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.KEYS;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.ORDERBY;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.SELECT_COLUMN_LIST;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.SELECT_PREFIX;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.SELECT_SUFFIX;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.TABLE;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.VALUES;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.WHERE;
+import static org.eclipse.dirigible.engine.odata2.sql.utils.OData2Utils.fqn;
+import static org.eclipse.dirigible.engine.odata2.sql.utils.OData2Utils.isOrderByEntityInExpand;
+
+import java.sql.Date;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.sql.Time;
+import java.sql.Timestamp;
+import java.util.ArrayList;
+import java.util.Calendar;
+import java.util.Collection;
+import java.util.HashSet;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.UUID;
+import java.util.stream.Collectors;
+
 import org.apache.olingo.odata2.api.commons.HttpStatusCodes;
-import org.apache.olingo.odata2.api.edm.*;
+import org.apache.olingo.odata2.api.edm.EdmEntitySet;
+import org.apache.olingo.odata2.api.edm.EdmEntityType;
+import org.apache.olingo.odata2.api.edm.EdmException;
+import org.apache.olingo.odata2.api.edm.EdmNavigationProperty;
+import org.apache.olingo.odata2.api.edm.EdmProperty;
+import org.apache.olingo.odata2.api.edm.EdmStructuralType;
+import org.apache.olingo.odata2.api.edm.EdmType;
 import org.apache.olingo.odata2.api.ep.entry.ODataEntry;
 import org.apache.olingo.odata2.api.exception.ODataException;
+import org.apache.olingo.odata2.api.exception.ODataHttpException;
 import org.apache.olingo.odata2.api.exception.ODataNotImplementedException;
 import org.apache.olingo.odata2.api.uri.NavigationPropertySegment;
 import org.apache.olingo.odata2.api.uri.NavigationSegment;
 import org.apache.olingo.odata2.api.uri.SelectItem;
-import org.apache.olingo.odata2.api.uri.expression.ExpressionKind;
+import org.apache.olingo.odata2.api.uri.UriInfo;
+import org.apache.olingo.odata2.api.uri.expression.CommonExpression;
 import org.apache.olingo.odata2.api.uri.expression.FilterExpression;
+import org.apache.olingo.odata2.api.uri.expression.MemberExpression;
 import org.apache.olingo.odata2.api.uri.expression.OrderByExpression;
 import org.apache.olingo.odata2.api.uri.expression.OrderExpression;
 import org.apache.olingo.odata2.core.edm.EdmDateTime;
@@ -31,24 +74,22 @@ import org.eclipse.dirigible.engine.odata2.sql.api.OData2Exception;
 import org.eclipse.dirigible.engine.odata2.sql.binding.EdmTableBinding;
 import org.eclipse.dirigible.engine.odata2.sql.binding.EdmTableBinding.ColumnInfo;
 import org.eclipse.dirigible.engine.odata2.sql.binding.EdmTableBindingProvider;
-import org.eclipse.dirigible.engine.odata2.sql.builder.expression.*;
+import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionDelete;
+import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionInsert;
+import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionJoin;
+import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionOrderBy;
+import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionSelect;
 import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionSelect.SQLSelectBuilder;
+import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionUpdate;
+import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionUtils;
+import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionWhere;
 import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionWhere.Param;
 import org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpressionWhere.TemporalType;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
-import java.sql.Date;
-import java.sql.*;
-import java.util.*;
-import java.util.stream.Collectors;
-
-import static org.apache.olingo.odata2.api.commons.HttpStatusCodes.BAD_REQUEST;
-import static org.apache.olingo.odata2.api.commons.HttpStatusCodes.INTERNAL_SERVER_ERROR;
-import static org.eclipse.dirigible.engine.odata2.sql.builder.expression.SQLExpression.ExpressionType.*;
-import static org.eclipse.dirigible.engine.odata2.sql.utils.OData2Utils.fqn;
-
 public final class SQLQuery {
+
     private static final Logger LOG = LoggerFactory.getLogger(SQLQuery.class);
 
     private final Map<String, EdmStructuralType> tableAliasesForEntitiesInQuery;
@@ -95,19 +136,46 @@ public final class SQLQuery {
         return this;
     }
 
-
     public SQLQuery orderBy(final OrderByExpression orderBy, final EdmEntityType entityType)
             throws EdmException, ODataNotImplementedException {
         if (orderBy != null && orderBy.getOrders() != null) {
             for (OrderExpression order : orderBy.getOrders()) {
-                if (ExpressionKind.PROPERTY != order.getExpression().getKind()) {
-                    LOG.error("OrderBy with non property expressions are not implemented yet!");
+                switch (order.getExpression().getKind()) {
+                case PROPERTY:
+                case MEMBER:                   
+                    continue;
+                default:
+                    LOG.error("OrderBy with non property or member expressions are not implemented yet!");
                     throw new ODataNotImplementedException();
                 }
             }
         }
         this.orderByExpressions = new SQLExpressionOrderBy(this, entityType, orderBy);
         return this;
+    }
+    
+    public void validateOrderBy(final UriInfo uriInfo)
+            throws EdmException, ODataNotImplementedException {
+        boolean usingOrderBy = uriInfo.getOrderBy() != null && uriInfo.getOrderBy().getOrders() != null;
+        if (usingOrderBy) {
+            for (OrderExpression orderExpression : uriInfo.getOrderBy().getOrders()) {
+                switch (orderExpression.getExpression().getKind()) {
+                case PROPERTY:
+                    continue;
+                case MEMBER:
+                    //For a member order by to work we need a corresponding expand (which also means a join).
+                    //Otherwise the table in the member order by clause will not be part of the query
+                    //E.g. SELECT T0.ID AS "ID_T0" FROM CARS AS T0 ORDER BY T1.FIRSTNAME DESC (we need a join to define T1)
+                    if (!isOrderByEntityInExpand(orderExpression, uriInfo)) {
+                        throw new OData2Exception("Missing $expand of the entity in the OrderBy expression", HttpStatusCodes.BAD_REQUEST); //no expand, but member order by is not allowed
+                    }
+                    continue;
+                default:
+                    LOG.error("OrderBy is implmented only for properties or member properties!");
+                    throw new ODataNotImplementedException();
+                }
+            }
+        }
     }
 
     public SQLQuery filter(final EdmEntitySet filterTarget, final FilterExpression filter) throws ODataException {
@@ -323,12 +391,6 @@ public final class SQLQuery {
     // This Method is for internal use ONLY !!! Do NEVER use it !!!
     // It will be hidden in future without further mitigation
     // TODO Refactor this method to private area
-    /**
-     * Get table alias
-     * 
-     * @param type the type
-     * @return the alias
-     */
     public String getSQLTableAlias(final EdmType type) {
         if (type instanceof EdmStructuralType)
             return getTableAliasForType((EdmStructuralType) type);
@@ -733,7 +795,6 @@ public final class SQLQuery {
         insertExpression.into(target, entry);
         return insertExpression;
     }
-
 
     public String buildInsert(final SQLContext context) throws EdmException, ODataException {
         StringBuilder builder = new StringBuilder();

--- a/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/builder/SQLQueryBuilder.java
+++ b/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/builder/SQLQueryBuilder.java
@@ -105,8 +105,9 @@ public class SQLQueryBuilder {
             q.filter(uri.getTargetEntitySet(), getKeyProperty(target), readIdsForExpand);
         }
         q.join(uri.getStartEntitySet(), uri.getTargetEntitySet(), uri.getNavigationSegments()).with(uri.getKeyPredicates());
+        q.validateOrderBy(uri);
         q.orderBy(uri.getOrderBy(), uri.getTargetEntitySet().getEntityType());
-
+        
         return q;
     }
 
@@ -132,7 +133,10 @@ public class SQLQueryBuilder {
         q.select(buildSelectItemsForPrimaryKey(target), null).top(effectiveTop).skip(effectiveSkip).from(target);
         q.filter(uri.getTargetEntitySet(), uri.getFilter())
                 .join(uri.getStartEntitySet(), uri.getTargetEntitySet(), uri.getNavigationSegments()).with(uri.getKeyPredicates());
-        q.orderBy(uri.getOrderBy(), uri.getTargetEntitySet().getEntityType());
+        //if the query is expanded, the order by is not necessary, because first the IDs will be selected
+        if (uri.getExpand().isEmpty()) {
+            q.orderBy(uri.getOrderBy(), uri.getTargetEntitySet().getEntityType());
+        }
 
         return q;
     }

--- a/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/builder/expression/SQLExpressionOrderBy.java
+++ b/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/builder/expression/SQLExpressionOrderBy.java
@@ -13,12 +13,16 @@ package org.eclipse.dirigible.engine.odata2.sql.builder.expression;
 
 import static org.apache.olingo.odata2.api.commons.HttpStatusCodes.INTERNAL_SERVER_ERROR;
 
+import java.util.ArrayList;
 import java.util.Iterator;
 import java.util.List;
 
 import org.apache.olingo.odata2.api.edm.EdmEntityType;
 import org.apache.olingo.odata2.api.edm.EdmException;
 import org.apache.olingo.odata2.api.edm.EdmProperty;
+import org.apache.olingo.odata2.api.edm.EdmStructuralType;
+import org.apache.olingo.odata2.api.uri.expression.CommonExpression;
+import org.apache.olingo.odata2.api.uri.expression.MemberExpression;
 import org.apache.olingo.odata2.api.uri.expression.OrderByExpression;
 import org.apache.olingo.odata2.api.uri.expression.OrderExpression;
 import org.apache.olingo.odata2.api.uri.expression.PropertyExpression;
@@ -26,18 +30,20 @@ import org.apache.olingo.odata2.api.uri.expression.SortOrder;
 import org.eclipse.dirigible.engine.odata2.sql.api.OData2Exception;
 import org.eclipse.dirigible.engine.odata2.sql.builder.SQLContext;
 import org.eclipse.dirigible.engine.odata2.sql.builder.SQLQuery;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
 
 public final class SQLExpressionOrderBy implements SQLExpression {
 
     private final OrderByExpression orderByExpression;
     private final SQLQuery query;
-    private final EdmEntityType orderByEnitityType;
+    private final EdmEntityType entityType;
+    private final Logger LOG = LoggerFactory.getLogger(SQLExpressionOrderBy.class);
 
     public SQLExpressionOrderBy(final SQLQuery query, final EdmEntityType orderByEnitityType, final OrderByExpression orderByExpression) {
-        //TODO extract the target enttiy from the orderByExpression
         this.orderByExpression = orderByExpression;
         this.query = query;
-        this.orderByEnitityType = orderByEnitityType;
+        this.entityType  = orderByEnitityType;
     }
 
     @Override
@@ -63,6 +69,7 @@ public final class SQLExpressionOrderBy implements SQLExpression {
 
     /**
      * This method parses the order by condition in the query.
+     * 
      * @param context
      * 
      * @param orderByExpression
@@ -70,40 +77,51 @@ public final class SQLExpressionOrderBy implements SQLExpression {
      * @throws ODataJPARuntimeException
      */
     private String parseExpression(SQLContext context) throws EdmException {
-        StringBuilder result = new StringBuilder();
+        List<String> orderByClauses = new ArrayList<>();
         if (!isEmpty()) {
             List<OrderExpression> orderBys = orderByExpression.getOrders();
-            String orderByField = null;
-            String orderByDirection = null;
-
             Iterator<OrderExpression> it = orderBys.iterator();
             while (it.hasNext()) {
                 OrderExpression orderBy = it.next();
-                try {
-                    EdmProperty prop = ((EdmProperty) ((PropertyExpression) orderBy.getExpression()).getEdmProperty());
-                    if (query.isTransientType(orderByEnitityType, prop)) {
-                        //unable to sort with a transient property in the list. This changes the semantic of order by and the result of the select
-                        throw new OData2Exception(
-                                "Unmapped property " + prop.getName()
-                                        + "! Unable to use an order by expresson for properties that are not mapped to the DB.",
-                                INTERNAL_SERVER_ERROR);
-                    }
-
-                    if (context == null || context.getDatabaseProduct() != null)
-                    	orderByField = query.getSQLTableColumn(orderByEnitityType, prop);
-                    else
-                    	orderByField = query.getSQLTableColumnAlias(orderByEnitityType, prop); // this gives the correct "order by" column name for Open SQL // TODO: add new enum type DatabaseProduct for OpenSQL
-
-                    orderByDirection = (orderBy.getSortOrder() == SortOrder.asc) ? "" : " DESC"; //$NON-NLS-1$
-                    result.append(orderByField).append(orderByDirection);
-                    if (it.hasNext()) {
-                        result.append(", ");
-                    }
-                } catch (EdmException e) {
-                    throw new OData2Exception("Unable to create ORDER BY clause: " + e.getMessage(), INTERNAL_SERVER_ERROR, e);
-                }
+                orderByClauses.add(orderByClause(context, orderBy));
             }
         }
-        return result.toString();
+        return SQLExpressionUtils.csv(orderByClauses);
+    }
+
+    protected String orderByClause(SQLContext context, OrderExpression orderBy) throws EdmException {
+        CommonExpression expression = orderBy.getExpression();
+        EdmStructuralType entityType = null;
+        EdmProperty prop = null;
+        StringBuilder orderByClause = new StringBuilder();
+
+        if (expression instanceof MemberExpression) {
+            MemberExpression memberExpression = (MemberExpression) expression;
+            CommonExpression pathExpression = memberExpression.getPath();
+            entityType = (EdmStructuralType) pathExpression.getEdmType();
+            PropertyExpression propertyExpression = (PropertyExpression) memberExpression.getProperty();
+            prop = (EdmProperty) propertyExpression.getEdmProperty();
+            
+        } else if (expression instanceof PropertyExpression) {
+            PropertyExpression propertyExpression = (PropertyExpression) expression;
+            prop = (EdmProperty) propertyExpression.getEdmProperty();
+            entityType = this.entityType;
+        } else {
+            throw new OData2Exception("Not Implemented", INTERNAL_SERVER_ERROR);
+        }
+        if (query.isTransientType(entityType, prop)) {
+            //unable to sort with a transient property in the list. This changes the semantic of order by and the result of the select
+            LOG.error("Unmapped property {}! Unable to use an order by expresson for properties that are not mapped to the DB.",
+                    prop.getName());
+            throw new OData2Exception(INTERNAL_SERVER_ERROR);
+        }
+
+        if (context == null || context.getDatabaseProduct() != null) {
+            orderByClause.append(query.getSQLTableColumn(entityType, prop));
+        } else {
+            orderByClause.append(query.getSQLTableColumnAlias(entityType, prop)); // this gives the correct "order by" column name for Open SQL // TODO: add new enum type DatabaseProduct for OpenSQL
+        }
+        orderByClause.append(" ").append(orderBy.getSortOrder() == SortOrder.asc ? "ASC" : "DESC");
+        return orderByClause.toString();
     }
 }

--- a/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/builder/expression/SQLExpressionUtils.java
+++ b/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/builder/expression/SQLExpressionUtils.java
@@ -82,7 +82,7 @@ public final class SQLExpressionUtils {
 	}
 	
 	public static String csv(List<String> values) {
-		return values.stream().collect(Collectors.joining(","));
+		return values.stream().collect(Collectors.joining(", "));
 	}
 
 }

--- a/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/processor/AbstractSQLProcessor.java
+++ b/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/processor/AbstractSQLProcessor.java
@@ -11,8 +11,30 @@
  */
 package org.eclipse.dirigible.engine.odata2.sql.processor;
 
+import static org.eclipse.dirigible.engine.odata2.sql.builder.EdmUtils.getProperties;
+import static org.eclipse.dirigible.engine.odata2.sql.builder.EdmUtils.getSelectedProperties;
+import static org.eclipse.dirigible.engine.odata2.sql.utils.OData2Utils.fqn;
+import static org.eclipse.dirigible.engine.odata2.sql.utils.OData2Utils.hasExpand;
+
+import java.io.InputStream;
+import java.sql.Connection;
+import java.sql.PreparedStatement;
+import java.sql.ResultSet;
+import java.sql.SQLException;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
 import org.apache.olingo.odata2.api.commons.InlineCount;
-import org.apache.olingo.odata2.api.edm.*;
+import org.apache.olingo.odata2.api.edm.EdmEntitySet;
+import org.apache.olingo.odata2.api.edm.EdmEntityType;
+import org.apache.olingo.odata2.api.edm.EdmException;
+import org.apache.olingo.odata2.api.edm.EdmLiteralKind;
+import org.apache.olingo.odata2.api.edm.EdmProperty;
+import org.apache.olingo.odata2.api.edm.EdmSimpleType;
+import org.apache.olingo.odata2.api.edm.EdmStructuralType;
 import org.apache.olingo.odata2.api.ep.EntityProvider;
 import org.apache.olingo.odata2.api.ep.EntityProviderReadProperties;
 import org.apache.olingo.odata2.api.ep.EntityProviderWriteProperties;
@@ -27,7 +49,16 @@ import org.apache.olingo.odata2.api.processor.ODataSingleProcessor;
 import org.apache.olingo.odata2.api.uri.KeyPredicate;
 import org.apache.olingo.odata2.api.uri.NavigationPropertySegment;
 import org.apache.olingo.odata2.api.uri.UriInfo;
-import org.apache.olingo.odata2.api.uri.info.*;
+import org.apache.olingo.odata2.api.uri.info.DeleteUriInfo;
+import org.apache.olingo.odata2.api.uri.info.GetComplexPropertyUriInfo;
+import org.apache.olingo.odata2.api.uri.info.GetEntitySetCountUriInfo;
+import org.apache.olingo.odata2.api.uri.info.GetEntitySetLinksUriInfo;
+import org.apache.olingo.odata2.api.uri.info.GetEntitySetUriInfo;
+import org.apache.olingo.odata2.api.uri.info.GetEntityUriInfo;
+import org.apache.olingo.odata2.api.uri.info.GetMediaResourceUriInfo;
+import org.apache.olingo.odata2.api.uri.info.GetSimplePropertyUriInfo;
+import org.apache.olingo.odata2.api.uri.info.PostUriInfo;
+import org.apache.olingo.odata2.api.uri.info.PutMergePatchUriInfo;
 import org.apache.olingo.odata2.core.uri.KeyPredicateImpl;
 import org.apache.olingo.odata2.core.uri.UriInfoImpl;
 import org.eclipse.dirigible.engine.odata2.sql.api.OData2EventHandler;
@@ -41,18 +72,6 @@ import org.eclipse.dirigible.engine.odata2.sql.utils.OData2ResultSetEntity;
 import org.eclipse.dirigible.engine.odata2.sql.utils.OData2Utils;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
-
-import java.io.InputStream;
-import java.sql.Connection;
-import java.sql.PreparedStatement;
-import java.sql.ResultSet;
-import java.sql.SQLException;
-import java.util.*;
-
-import static org.eclipse.dirigible.engine.odata2.sql.builder.EdmUtils.getSelectedProperties;
-import static org.eclipse.dirigible.engine.odata2.sql.builder.EdmUtils.getProperties;
-import static org.eclipse.dirigible.engine.odata2.sql.utils.OData2Utils.fqn;
-import static org.eclipse.dirigible.engine.odata2.sql.utils.OData2Utils.hasExpand;
 
 public abstract class AbstractSQLProcessor extends ODataSingleProcessor implements SQLProcessor {
 
@@ -294,7 +313,6 @@ public abstract class AbstractSQLProcessor extends ODataSingleProcessor implemen
 			LOG.error("Unable to serve request", e);
 			throw new ODataException(e);
 		}
-
 		return OData2Utils.writeFeedWithExpand(getContext(), (UriInfo) uriInfo, targetEntitiesResult, contentType,
 				count, nextLink);
 	}

--- a/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/utils/OData2Utils.java
+++ b/modules/odata/odata-core/src/main/java/org/eclipse/dirigible/engine/odata2/sql/utils/OData2Utils.java
@@ -11,8 +11,6 @@
  */
 package org.eclipse.dirigible.engine.odata2.sql.utils;
 
-import java.io.Closeable;
-import java.io.IOException;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
@@ -45,12 +43,12 @@ import org.apache.olingo.odata2.api.uri.ExpandSelectTreeNode;
 import org.apache.olingo.odata2.api.uri.NavigationPropertySegment;
 import org.apache.olingo.odata2.api.uri.UriInfo;
 import org.apache.olingo.odata2.api.uri.UriParser;
+import org.apache.olingo.odata2.api.uri.expression.CommonExpression;
+import org.apache.olingo.odata2.api.uri.expression.MemberExpression;
+import org.apache.olingo.odata2.api.uri.expression.OrderExpression;
 import org.eclipse.dirigible.engine.odata2.sql.api.OData2Exception;
-import org.slf4j.Logger;
-import org.slf4j.LoggerFactory;
 
 public class OData2Utils {
-    private static final Logger LOG = LoggerFactory.getLogger(OData2Utils.class);
 
     private OData2Utils() {
     }
@@ -280,6 +278,25 @@ public class OData2Utils {
             return false;
         }
         return true;
+    }
+
+
+    public static boolean isOrderByEntityInExpand(OrderExpression orderExpression, UriInfo uriInfo) throws EdmException {
+        MemberExpression memberExpression = (MemberExpression)orderExpression.getExpression();
+        CommonExpression pathExpression = memberExpression.getPath();
+        EdmType  entityType = pathExpression.getEdmType();
+        if (!hasExpand(uriInfo.getExpand())) {
+            return false;
+        }
+        for (List<NavigationPropertySegment> expand: uriInfo.getExpand()) {
+            for (NavigationPropertySegment segment: expand) {
+                EdmEntitySet epxandEntity = segment.getTargetEntitySet();
+                if (epxandEntity.getEntityType().getName().equals(entityType.getName())) {
+                    return true;
+                }
+             }
+        }
+        return false;
     }
 
 

--- a/modules/odata/odata-core/src/test/java/org/eclipse/dirigible/engine/odata2/sql/ODataSQLProcessorTest.java
+++ b/modules/odata/odata-core/src/test/java/org/eclipse/dirigible/engine/odata2/sql/ODataSQLProcessorTest.java
@@ -586,7 +586,7 @@ public class ODataSQLProcessorTest {
         List<ODataEntry> entries = resultFeed.getEntries();
         assertEquals(6, entries.size());
     }
-
+    
     @Test
     public void testExpand() throws Exception {
         Response response = OData2RequestBuilder.createRequest(sf) //
@@ -614,7 +614,94 @@ public class ODataSQLProcessorTest {
         Map<String, Object> secondDriverProperties = drivers.get(1).getProperties();
         assertEquals("Natalie", secondDriverProperties.get("FirstName"));
     }
+    
+    
+    @Test
+    public void testOrderByExpandedEntityWithoutExpand() throws Exception {
+        Response response = OData2RequestBuilder.createRequest(sf) //
+                .segments("Cars") //
+                .param("$top", "3") //
+                .param("$orderby", "Drivers/FirstName desc") //
+                //expand missing
+                .accept("application/atom+xml").executeRequest(GET);
+        assertEquals(500, response.getStatus()); //TODO refine the error codes, here it should be a bad request
+        
+    }
 
+    @Test
+    public void testOrderByExpandedEntityProperty() throws Exception {
+        Response response = OData2RequestBuilder.createRequest(sf) //
+                .segments("Cars") //
+                .param("$top", "3") //
+                .param("$orderby", "Drivers/FirstName desc") //
+                .param("$expand", "Drivers") //
+                .accept("application/atom+xml").executeRequest(GET);
+        assertEquals(200, response.getStatus());
+        
+        ODataFeed resultFeed = retrieveODataFeed(response, "Cars");
+        List<ODataEntry> entries = resultFeed.getEntries();
+        assertEquals("The limit must work with expand", 3, entries.size());
+        Map<String, Object> firstEntryProperties = entries.get(0).getProperties();
+        assertEquals(1982, firstEntryProperties.get("Year"));
+        assertEquals("Moskvitch", firstEntryProperties.get("Make"));
+        assertEquals("412", firstEntryProperties.get("Model"));
+
+        List<ODataEntry> drivers = ((ODataDeltaFeedImpl) (firstEntryProperties.get("Drivers"))).getEntries();
+        assertEquals(2, drivers.size());
+        Map<String, Object> firstDriverProperties = drivers.get(0).getProperties();
+        assertEquals("Natalie", firstDriverProperties.get("FirstName"));
+        Map<String, Object> secondDriverProperties = drivers.get(1).getProperties();
+        assertEquals("Johnny", secondDriverProperties.get("FirstName"));
+        
+        
+        
+        Response responseAsc = OData2RequestBuilder.createRequest(sf) //
+                .segments("Cars") //
+                .param("$top", "3") //
+                .param("$orderby", "Drivers/FirstName asc") //
+                .param("$expand", "Drivers") //
+                .accept("application/atom+xml").executeRequest(GET);
+        assertEquals(200, responseAsc.getStatus());
+        
+        ODataFeed resultFeedAsc = retrieveODataFeed(responseAsc, "Cars");
+        List<ODataEntry> entriesAsc = resultFeedAsc.getEntries();
+        assertEquals("The limit must work with expand", 3, entriesAsc.size());
+        Map<String, Object> firstEntryPropertiesAsc = entriesAsc.get(0).getProperties();
+        assertEquals(2015, firstEntryPropertiesAsc.get("Year"));
+        assertEquals("Ford", firstEntryPropertiesAsc.get("Make"));
+        assertEquals("S-Max", firstEntryPropertiesAsc.get("Model"));
+
+        List<ODataEntry> driversAsc = ((ODataDeltaFeedImpl) (firstEntryPropertiesAsc.get("Drivers"))).getEntries();
+        assertEquals(0, driversAsc.size());
+    }
+    
+    @Test
+    public void testOrderByExpandedEntityProperties() throws Exception {
+        Response response = OData2RequestBuilder.createRequest(sf) //
+                .segments("Cars") //
+                .param("$top", "3") //
+                .param("$orderby", "Drivers/FirstName desc,Drivers/LastName asc") //
+                .param("$expand", "Drivers") //
+                .accept("application/atom+xml").executeRequest(GET);
+        assertEquals(200, response.getStatus());
+        ODataFeed resultFeed = retrieveODataFeed(response, "Cars");
+        List<ODataEntry> entries = resultFeed.getEntries();
+        assertEquals("The limit must work with expand", 3, entries.size());
+        Map<String, Object> firstEntryProperties = entries.get(0).getProperties();
+        assertEquals(1982, firstEntryProperties.get("Year"));
+        assertEquals("Moskvitch", firstEntryProperties.get("Make"));
+        assertEquals("412", firstEntryProperties.get("Model"));
+
+        List<ODataEntry> drivers = ((ODataDeltaFeedImpl) (firstEntryProperties.get("Drivers"))).getEntries();
+        assertEquals(2, drivers.size());
+        Map<String, Object> firstDriverProperties = drivers.get(0).getProperties();
+        assertEquals("Natalie", firstDriverProperties.get("FirstName"));
+        Map<String, Object> secondDriverProperties = drivers.get(1).getProperties();
+        assertEquals("Johnny", secondDriverProperties.get("FirstName"));
+
+    }
+    
+    
     OData2RequestBuilder modifyingRequestBuilder(ODataServiceFactory sf, String content) {
         OData2RequestBuilder builder = new OData2RequestBuilder() {
             @Override

--- a/modules/odata/odata-core/src/test/java/org/eclipse/dirigible/engine/odata2/sql/builder/SQLQueryTest.java
+++ b/modules/odata/odata-core/src/test/java/org/eclipse/dirigible/engine/odata2/sql/builder/SQLQueryTest.java
@@ -173,7 +173,7 @@ public class SQLQueryTest {
         assertEquals("SELECT T0.MESSAGEGUID AS \"MESSAGEGUID_T0\", T0.LOGSTART AS \"LOGSTART_T0\", T0.LOGEND AS \"LOGEND_T0\", " +
                 "T0.SENDER AS \"SENDER_T0\", T0.RECEIVER AS \"RECEIVER_T0\", T0.STATUS AS \"STATUS_T0\", T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" " +
                 "FROM MPLHEADER AS T0 " +
-                "ORDER BY T0.STATUS, T0.LOGSTART DESC"
+                "ORDER BY T0.STATUS ASC, T0.LOGSTART DESC"
                 + SERVER_SIDE_PAGING_DEFAULT_SUFFIX, q.buildSelect(context));
     }
 
@@ -208,7 +208,7 @@ public class SQLQueryTest {
         assertEquals(
                 "SELECT T0.STATUS AS \"STATUS_T0\", T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" " +
                         "FROM MPLHEADER AS T0 " +
-                        "ORDER BY T0.STATUS, T0.LOGSTART DESC"
+                        "ORDER BY T0.STATUS ASC, T0.LOGSTART DESC"
                         + SERVER_SIDE_PAGING_DEFAULT_SUFFIX,
                 q.buildSelect(context));
     }
@@ -224,7 +224,7 @@ public class SQLQueryTest {
         SQLQuery q = builder.buildSelectEntitySetQuery(uriInfo);
 
         //The primary key is always selected in addition
-        assertEquals("SELECT T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" FROM MPLHEADER AS T0 ORDER BY T0.STATUS, T0.LOGSTART DESC"
+        assertEquals("SELECT T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" FROM MPLHEADER AS T0 ORDER BY T0.STATUS ASC, T0.LOGSTART DESC"
                 + SERVER_SIDE_PAGING_DEFAULT_SUFFIX, q.buildSelect(context));
     }
 
@@ -240,7 +240,7 @@ public class SQLQueryTest {
 
         //The primary key is always selected in addition
         assertEquals(
-                "SELECT T0.STATUS AS \"STATUS_T0\", T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" FROM MPLHEADER AS T0 ORDER BY T0.STATUS, T0.LOGSTART DESC"
+                "SELECT T0.STATUS AS \"STATUS_T0\", T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" FROM MPLHEADER AS T0 ORDER BY T0.STATUS ASC, T0.LOGSTART DESC"
                         + SERVER_SIDE_PAGING_DEFAULT_SUFFIX,
                 q.buildSelect(context));
     }
@@ -258,7 +258,7 @@ public class SQLQueryTest {
 
         //We expect to have FETCH FIRST expression with derby
         assertEquals("SELECT T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" " +
-                "FROM MPLHEADER AS T0 ORDER BY T0.STATUS, T0.LOGSTART " +
+                "FROM MPLHEADER AS T0 ORDER BY T0.STATUS ASC, T0.LOGSTART " +
                 "DESC FETCH FIRST 2 ROWS ONLY", q.buildSelect(context));
     }
 
@@ -275,7 +275,7 @@ public class SQLQueryTest {
         SQLContext context = new SQLContext(DatabaseProduct.POSTGRE_SQL);
 
         //We expect to have FETCH FIRST expression with derby
-        assertEquals("SELECT T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" FROM MPLHEADER AS T0 ORDER BY T0.STATUS, T0.LOGSTART DESC LIMIT 2",
+        assertEquals("SELECT T0.MESSAGEGUID AS \"MESSAGEGUID_T0\" FROM MPLHEADER AS T0 ORDER BY T0.STATUS ASC, T0.LOGSTART DESC LIMIT 2",
                 q.buildSelect(context));
     }
 

--- a/modules/odata/odata-core/src/test/java/org/eclipse/dirigible/engine/odata2/sql/builder/expression/SQLExpressionOrderByTest.java
+++ b/modules/odata/odata-core/src/test/java/org/eclipse/dirigible/engine/odata2/sql/builder/expression/SQLExpressionOrderByTest.java
@@ -51,19 +51,19 @@ public class SQLExpressionOrderByTest {
     @Test
     public void testSimpleOrderBy() throws Exception {
         SQLExpressionOrderBy sqlOrderBy = createOrderByExpression("Status");
-        assertEquals("T0.STATUS", sqlOrderBy.evaluate(null, ExpressionType.ORDERBY));
+        assertEquals("T0.STATUS ASC", sqlOrderBy.evaluate(null, ExpressionType.ORDERBY));
     }
 
     @Test
     public void testOrderByWith2Cols() throws Exception {
         SQLExpressionOrderBy sqlOrderBy = createOrderByExpression("Status, LogStart desc");
-        assertEquals("T0.STATUS, T0.LOGSTART DESC", sqlOrderBy.evaluate(null, ExpressionType.ORDERBY));
+        assertEquals("T0.STATUS ASC, T0.LOGSTART DESC", sqlOrderBy.evaluate(null, ExpressionType.ORDERBY));
     }
 
     @Test
     public void testOrderByWith3Cols() throws Exception {
         SQLExpressionOrderBy sqlOrderBy = createOrderByExpression("Status desc, LogEnd, LogStart desc");
-        assertEquals("T0.STATUS DESC, T0.LOGEND, T0.LOGSTART DESC", sqlOrderBy.evaluate(null, ExpressionType.ORDERBY));
+        assertEquals("T0.STATUS DESC, T0.LOGEND ASC, T0.LOGSTART DESC", sqlOrderBy.evaluate(null, ExpressionType.ORDERBY));
     }
 
     private SQLExpressionOrderBy createOrderByExpression(final String expression) {

